
#################################################################

# Location of the raw genomes
#RAW_GENOMES_PATH = "/mnt/moria/rylan/phenores_data/raw/genomes/"
RAW_GENOMES_PATH = "grdi_genomes/raw/"
#RAW_GENOMES_PATH = "test/genomes/"

# Kmer length that you want to count
KMER_SIZE = 11

# Data type of the resulting kmer matrix. Use uint8 if counts are
# all under 256. Else use uint16 (kmer counts under 65536)
MATRIX_DTYPE = 'uint8'

#################################################################

ids, = glob_wildcards(RAW_GENOMES_PATH+"{id}.fasta")

rule all:
  input:
    "grdi_touchfile.txt"

rule clean:
  input:
    RAW_GENOMES_PATH+"{id}.fasta"
  output:
    "grdi_genomes/clean/{id}.fasta"
  run:
    shell("python scripts/clean.py {input} grdi_genomes/clean/")

rule count:
  input:
    "grdi_genomes/clean/{id}.fasta"
  output:
    temp("grdi_jellyfish_results/{id}.jf")
  threads:
    2
  shell:
    "jellyfish count -C -m {KMER_SIZE} -s 100M -t {threads} {input} -o {output}"

rule dump:
    input:
        "grdi_jellyfish_results/{id}.jf"
    output:
        "grdi_jellyfish_results/{id}.fa"
    shell:
        "jellyfish dump {input} > {output}"

rule matrix:
    input:
        expand("grdi_jellyfish_results/{id}.fa", id=ids)
    output:
        touch("grdi_touchfile.txt")
    shell:
        "python scripts/parallel_matrix.py {KMER_SIZE} {MATRIX_DTYPE} grdi_jellyfish_results/ grdi_unfiltered/"
